lib/c-api/src/wasm_c_api/unstable/middlewares/metering.rs

```diff
@@ -0,0 +1,318 @@
+//!     assert(wasm_extern_kind(exports.data[0]) == WASM_EXTERN_FUNC);
+//!
+//!     const wasm_func_t* add_two = wasm_extern_as_func(exports.data[0]);
+//!     assert(add_two);
+//!
+//!     wasm_val_t arguments[1] = { WASM_I32_VAL(41) };
+//!     wasm_val_t results[1] = { WASM_INIT_VAL };
+//!
+//!     wasm_val_vec_t arguments_as_array = WASM_ARRAY_VEC(arguments);
+//!     wasm_val_vec_t results_as_array = WASM_ARRAY_VEC(results);
+//!
+//!     // Let's call `add_two` for the first time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         assert(trap == NULL);
+//!         assert(results[0].of.i32 == 42);
+//!
+//!         // There is 6 points left!
+//!         assert(wasmer_metering_get_remaining_points(instance) == 6);
+//!         assert(wasmer_metering_points_are_exhausted(instance) == false);
+//!     }
+//!
+//!     // Let's call `add_two` for the second time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         assert(trap == NULL);
+//!         assert(results[0].of.i32 == 42);
+//!
+//!         // There is 2 points left!
+//!         assert(wasmer_metering_get_remaining_points(instance) == 2);
+//!         assert(wasmer_metering_points_are_exhausted(instance) == false);
+//!     }
+//!
+//!     // Let's call `add_two` for the third time!
+//!     {
+//!         wasm_trap_t* trap = wasm_func_call(add_two, &arguments_as_array, &results_as_array);
+//!         // Oh, it failed!
+//!         assert(trap != NULL);
+//!
+//!         // There is 0 point left… they are exhausted.
+//!         assert(wasmer_metering_points_are_exhausted(instance) == true);
+//!     }
+//!     
+//!     wasm_instance_delete(instance);
+//!     wasm_module_delete(module);
+//!     wasm_store_delete(store);
+//!     wasm_engine_delete(engine);
+//!
+//!     return 0;
+//! }
+//! #    })
+//! #    .success();
+//! # }
+//! ```
+
+use super::super::super::instance::wasm_instance_t;
+use super::super::parser::operator::wasmer_parser_operator_t;
+use super::wasmer_middleware_t;
+use std::sync::Arc;
+use wasmer::wasmparser::Operator;
+use wasmer_middlewares::{
+    metering::{get_remaining_points, set_remaining_points, MeteringPoints},
+    Metering,
+};
+
+/// Opaque type representing a metering middleware.
+///
+/// To transform this specific middleware into a generic one, please
+/// see [`wasmer_metering_as_middleware`].
+///
+/// # Example
+///
+/// See module's documentation.
+#[allow(non_camel_case_types)]
+pub struct wasmer_metering_t {
+    pub(crate) inner: Arc<Metering<Box<dyn Fn(&Operator) -> u64 + Send + Sync>>>,
+}
+
+/// Function type to represent a user-defined cost function
+/// implemented in C.
+///
+/// # Example
+///
+/// See module's documentation.
+#[allow(non_camel_case_types)]
+pub type wasmer_metering_cost_function_t =
+    extern "C" fn(wasm_operator: wasmer_parser_operator_t) -> u64;
```

Comment:
Technically should probably be an `unsafe` function, but it's a minor point here because we immediately wrap it anyways. Only really matters if it has greater contact with Rust code.